src/NuGet.Protocol.Core.v3/DownloadResourceV3.cs

```diff
@@ -18,33 +18,81 @@ public class DownloadResourceV3 : DownloadResource
     {
         private readonly RegistrationResourceV3 _regResource;
         private readonly HttpClient _client;
+        private readonly string _flatContainerBaseUrl;
 
+        /// <summary>
+        /// Download packages using the download url found in the registration resource.
+        /// </summary>
         public DownloadResourceV3(HttpClient client, RegistrationResourceV3 regResource)
+            : this(client)
         {
             if (client == null)
             {
-                throw new ArgumentNullException("client");
+                throw new ArgumentNullException(nameof(regResource));
             }
 
-            if (regResource == null)
+            _regResource = regResource;
+        }
+
+        /// <summary>
+        /// Download packages using the flat container resource.
+        /// </summary>
+        public DownloadResourceV3(HttpClient client, string flatContainerBaseUrl)
+            : this(client)
+        {
+            if (flatContainerBaseUrl == null)
             {
-                throw new ArgumentNullException("regResource");
+                throw new ArgumentNullException(nameof(flatContainerBaseUrl));
+            }
+
+            _flatContainerBaseUrl = flatContainerBaseUrl.TrimEnd('/');
```

Comment:
It comes from index.json. I normalize the ending since this code constructs urls, but I don't think it should go too much farther to massage the url into the correct form.